File: boj/Sort/BOJ_1431.py
# 백준 1431 시리얼 번호
# 21.02.07

import sys
from functools import cmp_to_key


def compare(a, b):
    if len(a) == len(b):
        sumA = 0
        sumB = 0
        for i in range(len(a)):
            if '0' <= a[i] <= '9':
                sumA += ord(a[i]) - ord('0')
        for j in range(len(b)):
            if '0' <= b[j] <= '9':
                sumB += ord(b[j]) - ord('0')
        
        if sumA == sumB:
            if a > b: return 1
            else: return -1
        elif sumA > sumB: return 1
        else: return -1
    elif len(a) > len(b): return 1
    else: return -1


n = int(input())

s = []
for i in range(n):
    s.append(sys.stdin.readline().rstrip())

s.sort(key=cmp_to_key(compare))

for i in s:
    print(i)

File: boj/Sort/BOJ_1431.cpp
// 백준 1431 시리얼 번호
// 21.02.08
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#include <string>
#include <cmath>
#define INF 2147000000

typedef long long ll;

using namespace std;

bool compare(string a, string b)
{
    if (a.length() == b.length())
    {
        int sumA = 0, sumB = 0;
        for (int i = 0; i < a.length(); ++i)
        {
            if (a[i] >= '0' && a[i] <= '9')
            {
                sumA += a[i] - '0';
            }
        }
        for (int i = 0; i < b.length(); ++i)
        {
            if (b[i] >= '0' && b[i] <= '9')
            {
                sumB += b[i] - '0';
            }
        }

        if (sumA == sumB)
        {
            return a < b;
        }
        else return sumA < sumB;
    }
    else return a.length() < b.length();
}

int main()
{
    int n; cin >> n;
    vector <string> v;
    for (int i = 0; i < n; ++i)
    {
        string s; cin >> s;
        v.push_back(s);
    }

    sort(v.begin(), v.end(), compare);

    for (int i = 0; i < n; ++i)
        cout << v[i] << "\n";

    return 0;
}